packages/frontend/core/src/pages/workspace/index.tsx

```diff
@@ -37,24 +38,104 @@ declare global {
 
 export const Component = (): ReactElement => {
   performanceRenderLogger.debug('WorkspaceLayout');
+  const { workspacesService } = useServices({
+    WorkspacesService,
+  });
 
   const params = useParams();
+  const location = useLocation();
 
-  const [showNotFound, setShowNotFound] = useState(false);
-  const workspacesService = useService(WorkspacesService);
-  const listLoading = useLiveData(workspacesService.list.isLoading$);
-  const workspaces = useLiveData(workspacesService.list.workspaces$);
+  // check if we are in detail doc route, if so, maybe render share page
+  const detailDocRoute = useMemo(() => {
+    const match = matchPath(
+      '/workspace/:workspaceId/:docId',
+      location.pathname
+    );
+    if (
+      match &&
+      match.params.docId &&
+      match.params.workspaceId &&
+      // TODO(eyhn): need a better way to check if it's a docId
+      viewRoutes.find(route => matchPath(route.path, '/' + match.params.docId))
+        ?.path === '/:pageId'
+    ) {
+      return {
+        docId: match.params.docId,
+        workspaceId: match.params.workspaceId,
+      };
+    } else {
+      return null;
+    }
+  }, [location.pathname]);
 
+  const [workspaceNotFound, setWorkspaceNotFound] = useState(false);
+  const listLoading = useLiveData(workspacesService.list.isRevalidating$);
+  const workspaces = useLiveData(workspacesService.list.workspaces$);
   const meta = useMemo(() => {
     return workspaces.find(({ id }) => id === params.workspaceId);
   }, [workspaces, params.workspaceId]);
 
-  const workspace = useWorkspace(meta);
-  const globalContext = useService(GlobalContextService).globalContext;
+  // if listLoading is false, we can show 404 page, otherwise we should show loading page.
+  useEffect(() => {
+    if (listLoading === false && meta === undefined) {
+      setWorkspaceNotFound(true);
+    }
+    if (meta) {
+      setWorkspaceNotFound(false);
+    }
+  }, [listLoading, meta, workspacesService]);
 
+  // if workspace is not found, we should revalidate in interval
   useEffect(() => {
-    workspacesService.list.revalidate();
-  }, [workspacesService]);
+    if (listLoading === false && meta === undefined) {
+      const timer = setInterval(
+        () => workspacesService.list.revalidate(),
+        5000
+      );
+      return () => clearInterval(timer);
+    }
+    return;
+  }, [listLoading, meta, workspaceNotFound, workspacesService]);
+
+  if (workspaceNotFound) {
+    if (
+      detailDocRoute /*  */ &&
+      environment.isBrowser /* only browser has share page */
+    ) {
+      return (
+        <SharePage
```

Comment:
Not sure but I think the rendering should be handled in the child route, not in the parent route for clarity.

Reply:
the routes of the workspace page and the share page are exactly the same,

but the workspace has a sidebar and a topbar that the share page does not have.

So for share page we cannot render the workspace route.

Reply:
I see 👌